File: static/src/api/srp.ts
import request from '@/utils/http'

// ─── SRP 配置 ───

/** 获取 SRP 配置 */
export function fetchSrpConfig() {
  return request.get<Api.Srp.SrpConfig>({
    url: '/api/v1/srp/config'
  })
}

/** 更新 SRP 配置 */
export function updateSrpConfig(data: Api.Srp.SrpConfig) {
  return request.put({
    url: '/api/v1/srp/config',
    data
  })
}

// ─── 舰船价格表 ───

/** 获取舰船价格列表 */
export function fetchShipPrices(keyword?: string) {
  return request.get<Api.Srp.ShipPrice[]>({
    url: '/api/v1/srp/prices',
    params: keyword ? { keyword } : undefined
  })
}

/** 新增或更新舰船价格 */
export function upsertShipPrice(data: Api.Srp.UpsertShipPriceParams) {
  return request.post<Api.Srp.ShipPrice>({
    url: '/api/v1/srp/prices',
    data
  })
}

/** 删除舰船价格 */
export function deleteShipPrice(id: number) {
  return request.del({
    url: `/api/v1/srp/prices/${id}`
  })
}

// ─── 补损申请（用户端）───

/** 提交补损申请 */
export function submitApplication(data: Api.Srp.SubmitApplicationParams) {
  return request.post<Api.Srp.Application>({
    url: '/api/v1/srp/applications',
    data
  })
}

/** 获取我的补损申请列表 */
export function fetchMyApplications(params?: Partial<Api.Common.CommonSearchParams>) {
  return request.get<Api.Srp.ApplicationList>({
    url: '/api/v1/srp/applications/me',
    params
  })
}

/** 获取舰队范围内可用的 KM 列表（快捷申请） */
export function fetchFleetKillmails(fleetId: string) {
  return request.get<Api.Srp.FleetKillmailItem[]>({
    url: `/api/v1/srp/killmails/fleet/${fleetId}`
  })
}

/** 获取当前用户所有人物的全部 KM 列表（不限舰队；可按 characterId 筛选） */
export function fetchMyKillmails(characterId?: number) {
  return request.get<Api.Srp.FleetKillmailItem[]>({
    url: '/api/v1/srp/killmails/me',
    params: characterId ? { character_id: characterId } : undefined
  })
}

/** 获取 KM 装配详情 */
export function fetchKillmailDetail(data: Api.Srp.KillmailDetailRequest) {
  return request.post<Api.Srp.KillmailDetailResponse>({
    url: '/api/v1/srp/killmails/detail',
    data
  })
}

// ─── 补损申请（管理端）───

/** 获取全部申请列表（管理端） */
export function fetchApplicationList(params?: Api.Srp.ApplicationSearchParams) {
  return request.get<Api.Srp.ApplicationList>({
    url: '/api/v1/srp/applications',
    params
  })
}

/** 获取单条申请详情 */
export function fetchApplicationDetail(id: number) {
  return request.get<Api.Srp.Application>({
    url: `/api/v1/srp/applications/${id}`
  })
}

/** 审批补损申请 */
export function reviewApplication(id: number, data: Api.Srp.ReviewParams) {
  return request.put<Api.Srp.Application>({
    url: `/api/v1/srp/applications/${id}/review`,
    data
  })
}

/** 发放补损 */
export function payoutApplication(id: number, data?: Api.Srp.PayoutParams) {
  return request.put<Api.Srp.PayoutResult>({
    url: `/api/v1/srp/applications/${id}/payout`,
    data: data ?? {}
  })
}

/** 以伏羲币批量发放全部已批准未发放的 SRP */
export function batchPayoutAsFuxiCoin() {
  return request.put<Api.Srp.BatchFuxiPayoutActionResult>({
    url: '/api/v1/srp/applications/fuxi-payout',
    data: {}
  })
}

/** 获取批量发放汇总 */
export function fetchBatchPayoutSummary() {
  return request.get<Api.Srp.BatchPayoutSummary[]>({
    url: '/api/v1/srp/applications/batch-payout-summary'
  })
}

/** 批量自动审批符合规则的待审批 SRP */
export function runFleetAutoApproval(data: Api.Srp.AutoApproveParams) {
  return request.put<Api.Srp.AutoApproveSummary>({
    url: '/api/v1/srp/applications/auto-approve',
    data
  })
}

/** 按用户批量发放 SRP */
export function batchPayoutByUser(userId: number) {
  return request.put<Api.Srp.BatchPayoutActionResult>({
    url: `/api/v1/srp/applications/users/${userId}/payout`,
    data: {}
  })
}

/** 通过 ESI 在客户端打开人物信息窗口 */
export function openInfoWindow(data: { character_id: number; target_id: number }) {
  return request.post({
    url: '/api/v1/srp/open-info-window',
    data
  })
}

File: server/internal/handler/srp.go
package handler

import (
	"amiya-eden/internal/middleware"
	"amiya-eden/internal/model"
	"amiya-eden/internal/repository"
	"amiya-eden/internal/service"
	"amiya-eden/pkg/response"
	"fmt"
	"strconv"

	"github.com/gin-gonic/gin"
)

// SrpHandler 补损 HTTP 处理器
type SrpHandler struct {
	svc           *service.SrpService
	sysConfigRepo *repository.SysConfigRepository
}

func NewSrpHandler() *SrpHandler {
	return &SrpHandler{
		svc:           service.NewSrpService(),
		sysConfigRepo: repository.NewSysConfigRepository(),
	}
}

// ─────────────────────────────────────────────
//  舰船价格表
// ─────────────────────────────────────────────

// ListShipPrices GET /srp/prices
func (h *SrpHandler) ListShipPrices(c *gin.Context) {
	keyword := c.Query("keyword")
	list, err := h.svc.ListShipPrices(keyword)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, list)
}

// UpsertShipPrice POST /srp/prices
func (h *SrpHandler) UpsertShipPrice(c *gin.Context) {
	var req service.UpsertShipPriceRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeParamError, "请求参数错误: "+err.Error())
		return
	}
	userID := middleware.GetUserID(c)
	p, err := h.svc.UpsertShipPrice(userID, &req)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, p)
}

// DeleteShipPrice DELETE /srp/prices/:id
func (h *SrpHandler) DeleteShipPrice(c *gin.Context) {
	id := requireUintID(c, "id", " ID")
	if id == 0 {
		return
	}
	if err := h.svc.DeleteShipPrice(id); err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, nil)
}

// ─────────────────────────────────────────────
//  SRP 配置（仅 admin）
// ─────────────────────────────────────────────

// SrpConfigResponse SRP 配置响应
type SrpConfigResponse struct {
	AmountLimit float64 `json:"amount_limit"` // SRP 职权单笔上限（ISK）
}

// UpdateSrpConfigRequest 更新 SRP 配置请求
type UpdateSrpConfigRequest struct {
	AmountLimit *float64 `json:"amount_limit" binding:"required,min=0"`
}

// GetSrpConfig GET /srp/config
func (h *SrpHandler) GetSrpConfig(c *gin.Context) {
	amountLimit := h.sysConfigRepo.GetFloat(
		model.SysConfigSRPAmountLimit,
		model.SysConfigDefaultSRPAmountLimit,
	)
	response.OK(c, SrpConfigResponse{AmountLimit: amountLimit})
}

// UpdateSrpConfig PUT /srp/config
func (h *SrpHandler) UpdateSrpConfig(c *gin.Context) {
	var req UpdateSrpConfigRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeParamError, "请求参数错误: "+err.Error())
		return
	}
	if err := h.sysConfigRepo.Set(
		model.SysConfigSRPAmountLimit,
		fmt.Sprintf("%g", *req.AmountLimit),
		"SRP 职权单笔审批/发放上限（ISK）",
	); err != nil {
		response.Fail(c, response.CodeBizError, "更新 SRP 配置失败")
		return
	}
	response.OK(c, nil)
}

// ─────────────────────────────────────────────
//  申请（用户端）
// ─────────────────────────────────────────────

// SubmitApplication POST /srp/applications
func (h *SrpHandler) SubmitApplication(c *gin.Context) {
	var req service.SubmitApplicationRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeParamError, "请求参数错误: "+err.Error())
		return
	}
	userID := middleware.GetUserID(c)
	app, err := h.svc.SubmitApplication(userID, &req)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, app)
}

// ListMyApplications GET /srp/applications/my
func (h *SrpHandler) ListMyApplications(c *gin.Context) {
	page, pageSize, err := parsePaginationQuery(c, 20, 100)
	if err != nil {
		response.Fail(c, response.CodeParamError, err.Error())
		return
	}
	userID := middleware.GetUserID(c)

	list, total, err := h.svc.ListMyApplications(userID, page, pageSize)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OKWithPage(c, list, total, page, pageSize)
}

// GetMyKillmails GET /srp/my-killmails?character_id=xxx
func (h *SrpHandler) GetMyKillmails(c *gin.Context) {
	userID := middleware.GetUserID(c)
	var characterID int64
	if cidStr := c.Query("character_id"); cidStr != "" {
		if cid, err := strconv.ParseInt(cidStr, 10, 64); err == nil {
			characterID = cid
		}
	}
	kms, err := h.svc.GetMyKillmails(userID, characterID)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, kms)
}

// GetFleetKillmails GET /srp/fleet-killmails?fleet_id=xxx
func (h *SrpHandler) GetFleetKillmails(c *gin.Context) {
	fleetID := c.Param("fleet_id")
	if fleetID == "" {
		response.Fail(c, response.CodeParamError, "缺少 fleet_id 参数")
		return
	}
	userID := middleware.GetUserID(c)
	kms, err := h.svc.GetFleetKillmails(userID, fleetID)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, kms)
}

// ─────────────────────────────────────────────
//  申请管理（管理端：srp/fc/admin）
// ─────────────────────────────────────────────

// ListApplications GET /srp/manage/applications
func (h *SrpHandler) ListApplications(c *gin.Context) {
	page, pageSize, err := parseLedgerPaginationQuery(c, 20)
	if err != nil {
		response.Fail(c, response.CodeParamError, err.Error())
		return
	}

	filter := repository.SrpApplicationFilter{
		Tab:          repository.SrpTabType(c.Query("tab")),
		ReviewStatus: c.Query("review_status"),
		PayoutStatus: c.Query("payout_status"),
		Keyword:      c.Query("keyword"),
	}
	if fleetID := c.Query("fleet_id"); fleetID != "" {
		filter.FleetID = &fleetID
	}
	if charIDStr := c.Query("character_id"); charIDStr != "" {
		if cid, err := strconv.ParseInt(charIDStr, 10, 64); err == nil {
			filter.CharacterID = &cid
		}
	}

	list, total, err := h.svc.ListApplications(page, pageSize, filter)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OKWithPage(c, list, total, page, pageSize)
}

// ListBatchPayoutSummary GET /srp/applications/batch-payout-summary
func (h *SrpHandler) ListBatchPayoutSummary(c *gin.Context) {
	list, err := h.svc.ListBatchPayoutSummary()
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, list)
}

// RunFleetAutoApproval PUT /srp/applications/auto-approve
func (h *SrpHandler) RunFleetAutoApproval(c *gin.Context) {
	var req service.RunFleetAutoApprovalRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeParamError, "请求参数错误: "+err.Error())
		return
	}
	reviewerID := middleware.GetUserID(c)
	callerRoles := middleware.GetUserRoles(c)
	result, err := h.svc.RunFleetAutoApproval(reviewerID, callerRoles, req.FleetID)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, result)
}

// GetApplication GET /srp/manage/applications/:id
func (h *SrpHandler) GetApplication(c *gin.Context) {
	id := requireUintID(c, "id", " ID")
	if id == 0 {
		return
	}
	app, err := h.svc.GetApplication(id)
	if err != nil {
		response.Fail(c, response.CodeNotFound, "申请不存在")
		return
	}
	response.OK(c, app)
}

// ReviewApplication PATCH /srp/manage/applications/:id/review
func (h *SrpHandler) ReviewApplication(c *gin.Context) {
	id := requireUintID(c, "id", " ID")
	if id == 0 {
		return
	}
	var req service.ReviewApplicationRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeParamError, "请求参数错误: "+err.Error())
		return
	}
	reviewerID := middleware.GetUserID(c)
	callerRoles := middleware.GetUserRoles(c)
	app, err := h.svc.ReviewApplication(reviewerID, callerRoles, id, &req)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, app)
}

// Payout PATCH /srp/manage/applications/:id/payout
func (h *SrpHandler) Payout(c *gin.Context) {
	id := requireUintID(c, "id", " ID")
	if id == 0 {
		return
	}
	var req service.SrpPayoutRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeParamError, "请求参数错误: "+err.Error())
		return
	}
	payerID := middleware.GetUserID(c)
	callerRoles := middleware.GetUserRoles(c)
	app, mailSummary, err := h.svc.Payout(payerID, callerRoles, id, &req)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, service.SrpPayoutActionResult{
		SrpApplication:     *app,
		MailAttemptSummary: mailSummary,
	})
}

// BatchPayoutByUser PUT /srp/applications/users/:user_id/payout
func (h *SrpHandler) BatchPayoutByUser(c *gin.Context) {
	userID := requireUintID(c, "user_id", "用户 ID")
	if userID == 0 {
		return
	}
	payerID := middleware.GetUserID(c)
	callerRoles := middleware.GetUserRoles(c)
	result, mailSummary, err := h.svc.BatchPayoutByUser(payerID, callerRoles, userID)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, service.SrpBatchPayoutActionResult{
		SrpBatchPayoutSummaryResponse: *result,
		MailAttemptSummary:            mailSummary,
	})
}

// BatchPayoutAsFuxiCoin PUT /srp/applications/fuxi-payout
func (h *SrpHandler) BatchPayoutAsFuxiCoin(c *gin.Context) {
	payerID := middleware.GetUserID(c)
	callerRoles := middleware.GetUserRoles(c)
	result, mailSummary, err := h.svc.BatchPayoutAsFuxiCoin(payerID, callerRoles)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, service.SrpBatchFuxiPayoutActionResult{
		SrpBatchFuxiPayoutSummary: *result,
		MailAttemptSummary:        mailSummary,
	})
}

// OpenInfoWindow POST /srp/open-info-window
// 通过 ESI 在客户端打开人物信息窗口
func (h *SrpHandler) OpenInfoWindow(c *gin.Context) {
	var req service.OpenInfoWindowRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeParamError, "请求参数错误: "+err.Error())
		return
	}
	userID := middleware.GetUserID(c)
	if err := h.svc.OpenInfoWindow(userID, &req); err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, nil)
}

// GetKillmailDetail POST /srp/killmails/detail
func (h *SrpHandler) GetKillmailDetail(c *gin.Context) {
	var req service.KillmailDetailRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Fail(c, response.CodeParamError, "请求参数错误: "+err.Error())
		return
	}
	detail, err := h.svc.GetKillmailDetail(&req)
	if err != nil {
		response.Fail(c, response.CodeBizError, err.Error())
		return
	}
	response.OK(c, detail)
}
